### PR TITLE
Fixed MM-18154: attach also searches for issue key

### PR DESCRIPTION
**Pull request by levb**
Re-implemented, server-side.

Fixes https://mattermost.atlassian.net/browse/MM-18154

### CODE REVIEW

File: webapp/src/components/jira_issue_selector/jira_issue_selector.jsx [hunk 1]
```diff
@@ -55,9 +55,14 @@ export default class JiraIssueSelector extends Component {
...
     searchIssues = (text) => {
-        const textEncoded = encodeURIComponent(text.trim().replace(/"/g, '\\"'));
-        const textSearchTerm = (textEncoded.length > 0) ? 'text ~ "' + textEncoded + '*"' : '';
-        const finalQuery = textSearchTerm + ' ' + searchDefaults;
+        const trimmed = text.trim();
+        if (trimmed === '') {
+            return {};
+        }
+        const textEncoded = encodeURIComponent(trimmed.replace(/"/g, '\\"'));
+        const textSearchTerm = 'text ~ "' + textEncoded + '*"';
```
Comment (jfrerich) [inline webapp/src/components/jira_issue_selector/jira_issue_selector.jsx]: `2/5` propose `s/textKeyTerm/textIssueKeyTerm/` so the reading constructed `finalQuery` variable is more obvious.

Comment (jfrerich) [reply webapp/src/components/jira_issue_selector/jira_issue_selector.jsx]: https://github.com/mattermost/mattermost-plugin-jira/pull/325/files#diff-7a8afa6c78a7c8d826b0b0a89d826925R65

Comment (levb) [reply webapp/src/components/jira_issue_selector/jira_issue_selector.jsx]: The better fix IMO is to move all this to the server, and re-define the API as "IssueAutocomplete", but I agree, will make this more intention-revealing.

File: webapp/src/components/jira_issue_selector/jira_issue_selector.jsx [hunk 2]
```diff
@@ -55,9 +55,14 @@ export default class JiraIssueSelector extends Component {
     };
 
     searchIssues = (text) => {
-        const textEncoded = encodeURIComponent(text.trim().replace(/"/g, '\\"'));
-        const textSearchTerm = (textEncoded.length > 0) ? 'text ~ "' + textEncoded + '*"' : '';
-        const finalQuery = textSearchTerm + ' ' + searchDefaults;
+        const trimmed = text.trim();
+        if (trimmed === '') {
+            return {};
```
Comment (mickmister) [inline webapp/src/components/jira_issue_selector/jira_issue_selector.jsx]: This function is supposed to return a `Promise<[]>`, so I suggest making the function async, and returning a `[]` here.

File: webapp/src/components/jira_issue_selector/jira_issue_selector.jsx (hunk 1 above)
Comment (mickmister) [reply webapp/src/components/jira_issue_selector/jira_issue_selector.jsx]: :+1: on moving to server

**Comment by DHaussermann** [comment]
@levb I'm not seeing this work as I expected. 

1. Searching by text across summary and description no longer works. Any string entered that does not line up with an issue key results in the error showing `Error: failed to get search results: - The issue key 'search text' for field 'issuekey' is invalid.`  If letter characters I'm typing don't line up with a project key, I would expect the old behavior.

Other UX issues:

2. The error is very eager to display. If issue key `DKH-32` exists, as soon as I type 'D` I get the error that it's not a valid issue key. Ideally this message should only show after I type the project key followed by a dash and then the 1st number. The error would then be shown if there are no possible matches.  
Issue 1. may already solve this if it can be changed to the old behavior until the string matches a project key. 
 
3. The auto-complete drop down only shows the one result that is an exact match. If issue key `DKH-32` exists and I type `DKH-3` I would expect to show issue 3 and the top but also show 30, 31, 32 as available options. Currently there is no feedback that the search returns multiple matches.    

Tested on Server. Testing on Cloud is ongoing.

**Comment by levb** [comment]
#1 - on it.

#2 - I'll see what I can do, I wanted to  leave the change itself small, but maybe the 404 error should not be displayed until the user typed at least some number of characters.

#3 - not easily possible, Jira does not support regex searches on issue keys.

**Comment by DHaussermann** [comment]
Added `Do Not Merge` label for clarity as there is a code change coming that will need review and testing.

File: server/issue.go [hunk 3]
```diff
@@ -363,32 +367,51 @@ func httpAPIGetSearchIssues(ji Instance, w http.ResponseWriter, r *http.Request)
...
-
-	searchRes, err := client.SearchIssues(jqlString, &jira.SearchOptions{
-		MaxResults: 50,
-		Fields:     []string{"key", "summary"},
-	})
-	if err != nil {
-		return http.StatusInternalServerError,
-			errors.WithMessage(err, "failed to get search results")
+	q := r.FormValue("q")
+	var exact *jira.Issue
```
Comment (jfrerich) [inline server/issue.go]: nit: Propose to change `exact` to `exactIssue` or something in the variable name that recognizes `exact` is of type `*jira.Issue`

Comment (levb) [reply server/issue.go]: That'd be too Java...

Comment (jfrerich) [reply server/issue.go]: Noted that Java naming convention is bad! 😆

**Pull request merged**